Add span watcher processor to SDK spec (#6)
dynatrace-oss-contrib/opentelemetry-specification: specification/sdk-tracing.md

* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

##### Span watcher processor

> This implementation passes spans that have been started but not yet ended

Reviewer: Not sure if it is a good idea to have a dedicated processor reporting only open spans. This more or less implies that an exporter has to be able to handle more parallel streams or several exporters run in parallel even if they have the same target.
I would prefer an processor which exports open and finished spans. What we need is a clear indication if a span is finished or not.

Reviewer 2: In fact, it might be worth considering if adding an option `exportActiveSpans` to the `BatchSpanProcessor` would be better than a dedicated span processor. I think an advantage would be reduced code duplication in addition to the better synchronization @Flarna mentions. The only disadvantage would be that this may add quite some overhead if the option is not used.

Reviewer 2: > What we need is a clear indication if a span is finished or not.

👍  This is absolutely required.

Author: I agree, this should make exporting easier. I'll add this as an option.

Author: I agree, this would make exporting easier. I'll add an option to have the span watcher processor also export ended spans for now. We can discuss merging it into the `BatchSpanProcessor` with the community later on.

to the configured exporter.

> On each interval defined by `reportIntervalMillis`, all unfinished spans that

Reviewer 2: Is "unfinished" a well-defined term? I always use "un-ended" because even though it does sound a bit awkward, it is at least very clear.

Author: You're right, "finished" could be interpreted as "done with exporting" as well, for example. I'll change it to "un-ended" until we find a less awkward term.

> have been started more than `reportIntervalMillis` ago, are exported as is.

Reviewer: This implies that one more, parallel export is started if the previous has not yet finished. This may result in a lot parallel exports on hanging connections/bad set intervals.

Author: Should we define the report interval as the duration between two exports (export1.end -> interval -> export2.start), i.e., define the export as blocking rather than having a timer that triggers a new export every n millis?
An alternative would be to specify that exports should be skipped if a previous export is still running at the next timer interval.

**Configurable parameters:**

* `exporter` - the exporter where the spans are pushed.

> * `maxQueueSize` - the maximum queue size. After the size is reached spans are

Reviewer: I think it should be specified which spans to drop (e.g. oldest, new ones added,...)

Author: Indeed, this should be specified for the batch processor as well.
What do you think would be best?
I can imagine use cases for oldest (long-running), newest (the ones rushing in recently, causing the overflow) as well as random ones (would cover both).

Reviewer: Easiest is to drop new spans. As there is no `correct` this should be fine.

> * `maxQueueSize` - the maximum queue size. After the size is reached spans are

Reviewer 2: Also, its not a queue. It's a list (I like to think of it as the "watchlist). A queue implies to me that all items in it are processed once and then removed from the queue.

  dropped. The default value is `2048`.
* `reportIntervalMillis` - the delay interval in milliseconds between two
  consecutive exports. The default value is `5000`.

> * `exporterTimeoutMillis` - how long the export can run before it is cancelled.

Reviewer: Exporters have no `cancel` operation therefore this configuration is not applicable for a SpanProcessor.

Author: I copied this one from the batching processor in the paragraph above. This part of the spec is not yet implemented in all languages repos yet, however.

  The default value is `30000`.
* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

For languages that support the concept of weak references it is recommended to
only use these for keeping references to unfinished spans in the processor.
Otherwise, it would accumulate references to spans abandoned by the user without
ending them and therefore potentially leak memory.
If weak references are not supported, a timeout parameter should be added for
this processor, after which unfinished spans are dropped without reporting
further updates. This timeout, however, should only apply to this processor and
MUST NOT affect spans that are ended properly after the timeout. Other
processors, including the simple and batching processor, will still export

> these spans once finished.

Reviewer 3: The batch span processor doesn't export when spans finish, but on some interval
```suggestion
these spans as they normally would.
```

> * `spanDurationTimeoutMillis` - no further updates are sent for spans that have

Reviewer 3: Is this intentionally separate from the other configurable parameters? 

Is the processor expected to drop spans as soon as they get too old, or should they wait until the next reporting interval?

For languages that do support weak references, does this configuration option still exist?

The wording on this feels weird to me. I think I would prefer something along the lines of "On each report interval, spans older than `spanDurationTimeoutMillis` which have not yet ended will be dropped." This makes it more clear when the drop will happen too.

Author: I would only add this option for languages that don't support weak references and therefore separated it from the general options. I can see that it looks a bit misplaced however and will move it to the others and add a note.

On the next interval makes more sense to me, I'll adopt your wording.

> * `spanDurationTimeoutMillis` - no further updates are sent for spans that have

Reviewer: Is `Span.startTime` relevant here or the time when `processor.onStart()` is called? Please note that it is allowed to pass a timestamp to `startSpan()` therefore these two points may differ a lot.

Author: I think both would be valid but we should decide for one. Which one would you find more reasonable?

Reviewer: Currently our cluster discards all traces whcih are out of time so as of now we can't support a significant drift anyway.
I would favor to use the time when processor get's the span otherwise it may drop it fast because it looks old.

Author: I added "_Unfinished spans which were first provided to the processor longer than this timeout ago..._"

* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

##### Span watcher processor

This implementation passes spans that have been started but not yet ended
to the configured exporter. These spans are called _unfinished_ spans.
On each interval defined by `reportIntervalMillis`, all unfinished spans that

> have been started more than `reportIntervalMillis` ago, are exported.

Reviewer 2: ```suggestion
have been started at least `reportIntervalMillis` ago, are exported.
```

**Configurable parameters:**

> * `exporter` - the exporter where the spans are pushed.

Reviewer 2: ```suggestion
* `exporter` - the exporter to which the spans are pushed.
```

* `maxWatchedSpans` - the maximum number of unfinished spans to be watched.
  While reached, no new spans will be accepted. The default value is `2048`.

> * `reportIntervalMillis` - the delay interval in milliseconds between two

Reviewer 2: ```suggestion
* `reportIntervalMillis` - the mimimum delay interval in milliseconds between two
```

  consecutive exports. The default value is `5000`.
* `exporterTimeoutMillis` - how long the export can run before it is cancelled.
  The default value is `30000`.
* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxWatchedSpans`. The default value is `512`.
* `exportEndedSpans` (boolean) - if set, spans are also exported once they are
  ended.
  The default value is `false`.

> * `spanDurationTimeoutMillis` - On each report interval, spans older than

Reviewer 2: ```suggestion
* `maxSpanDurationMillis` - On each report interval, spans older than
```

  `spanDurationTimeoutMillis` which have not yet ended will be dropped from
  the watch list.

>   The default is `60000`.

Reviewer 2: One minute? That seems a bit strict. Also, I would suggest adding

```suggestion
  The default is `60000` (1 minute).
```

  (_only applies to languages where weak references are not supported or
  suitable_)

In order to avoid multiple exports potentially being blocked on connection issues or

> when `reportIntervalMillis` is set inappropriately, exports should *not* be performed

Reviewer 2: But when it *is* set appropriately, they may be performed in parallel?

in parallel. If a previous export is still in progress on the next

> `reportIntervalMillis`, the current export should be skipped.

Reviewer 2: "on the next reportIntervalMillis" sounds strange to me. I suggest "If an export would normally happen while another export by this SpanProcessor is still running, the export is skipped."

For languages that support the concept of weak references it is recommended to
only use these for keeping references to unfinished spans in the processor.
Otherwise, it would accumulate references to spans abandoned by the user without
ending them and therefore potentially leak memory.
If weak references are not supported or suitable to be used, the timeout parameter
`maxSpanDurationMillis` should be added for this processor. Unfinished spans which

> were first provided to the processor longer than this timeout ago are dropped from the

Reviewer 2: ```suggestion
were first provided to the processor longer ago than this timeout are dropped from the
```

* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxQueueSize`. The default value is `512`.

##### Span watcher processor

This implementation passes spans that have been started but not yet ended
to the configured exporter.
On each interval defined by `reportIntervalMillis`, all un-ended spans that

> have been started more than `reportIntervalMillis` ago, are exported as is.

Reviewer 4: What does "as is" mean here?

Author: _as is_ as in _as they currently are_, i.e., their current state
that's actually redundant, however, since there's no other (future) state that can be exported anyway, so I removed it

**Configurable parameters:**

* `exporter` - the exporter where the spans are pushed.

> * `maxWatchedSpans` - the maximum number of un-ended spans to be watched.

Reviewer 4: The term "un-ended" feels really cumbersome to me. Let's add some initial terminology clarification -  e.g:

> Such spans that have been started but not yet ended will be referred to as "processing|open|... spans" from now on.

Author: I replaced it with "unfinished" and defined the terminology in the introduction. What do you think? _Open_ would be an option but I'd rather only use that if the start/end methods were called open and close.

  While reached, no new spans will be accepted. The default value is `2048`.
* `reportIntervalMillis` - the delay interval in milliseconds between two
  consecutive exports. The default value is `5000`.
* `exporterTimeoutMillis` - how long the export can run before it is cancelled.
  The default value is `30000`.
* `maxExportBatchSize` - the maximum batch size of every export. It must be
  smaller or equal to `maxWatchedSpans`. The default value is `512`.
* `exportEndedSpans` (boolean) - if set, spans are also exported once they are
  ended.

>   The default value is `false`.

Reviewer: I would use `true` as default value here as I assume most users are interested on finished spans also.

Author: This depends on the use case. If I use it for debugging purposes like the zPages feature of OpenCensus I will likely have a default exporter for finished spans already in place and will use the span watcher only for running spans.

Reviewer: Don't know what zPages are. But even for debugging I would like to have open and finished spans in one (e.g. log) exporter.

* `spanDurationTimeoutMillis` - On each report interval, spans older than
  `spanDurationTimeoutMillis` which have not yet ended will be dropped from
  the watch list.
  The default is `60000`.
  (_only applies to languages where weak references are not supported or
  suitable_)

In order to avoid multiple exports potentially being blocked on connection issues or
when `reportIntervalMillis` is set inappropriately, exports should *not* be performed
in parallel. If a previous export is still in progress on the next
`reportIntervalMillis`, the current export should be skipped.

For languages that support the concept of weak references it is recommended to
only use these for keeping references to unfinished spans in the processor.
Otherwise, it would accumulate references to spans abandoned by the user without
ending them and therefore potentially leak memory.
If weak references are not supported or suitable to be used, the timeout parameter

> `spanDurationTimeoutMillis` should be added for this processor. Unfinished spans which

Reviewer: ```suggestion
`spanDurationTimeoutMillis` should be added by this processor. Unfinished spans which
```

Author: To me this spec is an instruction for the person writing the processor implementation so I'd stick with _for_ (or _to_?).

`maxSpanDurationMillis` should be added for this processor. Unfinished spans which
were first provided to the processor longer than this timeout ago are dropped from the
list of watched spans and therefore updates will no longer be reported for them.
This timeout, however, should only apply to this processor and MUST NOT affect spans

> that are ended properly after the timeout. If `exportEndedSpans` is set, the span MUST

Reviewer 2: Oof, I need to update my implementation 😅  But makes sense. Should they also be exported if the span watchlist is full at the time they are ended? If yes, is there a dedicated limit for number of queued ended spans?

Author: I think both would be okay. It could either be implemented as two separate lists with two limits or one list with a common limit or again two limits. If using one list is easier/more efficient, having two limits would have the advantage that one can set maxEnded > maxWatched so that too many watched spans never suppress sending ended spans in a high load scenario.

Reviewer 2: This is actually a quite interesting question; how to mix ended and unfinished spans. I can see several possibilities:

(NOTE: If you use weak references, then you need to make sure you hold on to the spans in onEnd, which may be implemented by storing spans in a WeakMap instead of a simple List<WeakRef> and setting the value of WeakMap.get(endedSpan). This is an important consideration to estimate the ease of implementation of the variants below.).

1. Don't deal with ended spans at all in the SpanWatcher. Discard a span as soon as it is ended. Employ a separate BatchSpanProcessor for ended spans. HTTP frameworks hopefully deal with recycling the TCP connection, but separate HTTP requests will be made for unfinished and ended spans.
2. Deal with them in the SpanWatcher (as currently specified). Several variants:
   1. Simply send ended spans once, then remove them. If you have strong refs to spans anyway, you don't even need onEnd at all (see note above).
   2. As above, but with "ended span recovery". Consider the case that the watchlist was full at the time a span was started or the span has already timed out (older than maxAge). Then you would still get an onEnd callback in your span processor, which you can use to still potentially send the span. Again, several variants exists:
      1. Single list for ended and unfinished spans: Do in onEnd as you would in onStart, i.e. try to insert the span into the watchlist again. This requires being able to find the span in the watchlist / determining if it is there.
         1. Use the same limit for ended and unfinished spans.
         2. Use a separate limit (typically higher) that allows ended spans to be inserted into the list even if no unfinished would be watched anymore (i.e., reserverve part of the list for only ended spans).
       2. Separate list for ended and unfinished spans. This will require some way to avoid sending spans twice and seeing that spans can be ended at any point in time, might be difficult to do free of race-conditions. But I haven't thought this through.

Reviewer 2: I think I would prefer a 2.ii.a variant, but if someone can describe how to nicely implement 2.ii.b, that would also be fine.

Author: Thanks for the detailed write-up! Variant 2.ii.a.b sounds reasonable to me but either way I don't think that the spec should dictate the implementation in that detail. I will define two separate limits for the reason stated in 2.ii.a.b and leave the rest up to the SIGs.

Reviewer: I would not overcomplicate this. Once we reach a state that exporter is unable to export fast enough spans are lost and trace is damaged as references are missing. The behavior during such phases is mostly here to avoid more damage to the process. Trying to find a solution to get the best out of the remaining spans is most likely quite useless as the traces with missing spans are garbage.

> processor, will still export these spans as they normally would.

Reviewer: Consider to remove `Other processors, includi...` as it should be anyway clear that one processor should not impact others in any aspect.
